### PR TITLE
feat(exec): Explicit Docker image prefetch

### LINKED ISSUE #5197: lib/exec/docker: perform docker pull once per tag per run
Instead of letting "docker run" implicitly pull images, let's do a pre-emptive pull of the image/tag combination once we know it. This will also lead to more accurate timing of exec calls.

However "docker pull" usually takes a few seconds even when there's nothing to do , so we don't want to do it every time (especially for our setup.py extract calls). So instead, let's keep an array of images we've already pulled and not pull them again "per run"/boot.

### PR DESCRIPTION
**Pull request by zharinov**
Closes #5197

### CODE REVIEW

File: lib/util/exec/index.ts [hunk 1]
```diff
@@ -17,11 +17,14 @@ const execConfig: ExecConfig = {
...
 };
 
+const prefetchedDockerImages: Set<string> = new Set<string>();
+
 export function setExecConfig(config: Partial<RenovateConfig>): void {
   for (const key of Object.keys(execConfig)) {
     const value = config[key];
     execConfig[key] = value || null;
   }
+  prefetchedDockerImages.clear();
```
Comment (zharinov) [inline lib/util/exec/index.ts]: Should be good enough for independent test contexts, as soon as we call `setExecConfig` once per run

File: lib/util/exec/index.ts [hunk 2]
```diff
@@ -104,7 +107,13 @@ export async function exec(
     let singleCommand = commands.join(' && ');
     singleCommand = `bash -l -c "${singleCommand.replace(/"/g, '\\"')}"`;
     singleCommand = dockerCmd(singleCommand, dockerOptions, execConfig);
-    commands = [singleCommand];
+
+    const { image, tag = 'latest' } = docker;
```
Comment (rarkins) [inline lib/util/exec/index.ts]: Wouldn't this logic better belong inside util/exec/docker.ts ?

File: lib/util/exec/docker/index.ts [hunk 3]
```diff
@@ -68,5 +74,12 @@ export function dockerCmd(
...
 
-  return result.join(' ');
+  const mainCommand = result.join(' ');
+
+  if (prefetchedDockerImages.has(taggedImage)) {
+    return [mainCommand];
+  }
+
+  prefetchedDockerImages.add(taggedImage);
+  return [`docker pull ${taggedImage}`, mainCommand];
```
Comment (rarkins) [inline lib/util/exec/docker/index.ts]: No, let's pull it using a separate child_process. We don't want it to be counted as part of the time taken to exec

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/exec/docker/index.ts (hunk 3 above)
Comment (zharinov) [reply lib/util/exec/docker/index.ts]: Let's merge #5207, then I'll deconflict & update it

Comment (rarkins) [reply lib/util/exec/docker/index.ts]: ok. I'm having test case failures for some reason, hope to resolve soon

Comment (zharinov) [reply lib/util/exec/docker/index.ts]: Can be fixed by `join('...')` from `upath`, if it's Windows CI failure

Comment (rarkins) [reply lib/util/exec/docker/index.ts]: I skipped it, updated this branch from master

**Comment by zharinov** [comment]
Ready for review

**Pull request merged**